crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+
+    fn get_committee_by_epoch(&self, epoch: EpochId) -> Option<Committee>;
+}
+
+pub trait CommitteeStore {
+    fn insert_committee(&self, new_committee: &Committee);
+
+    fn get_committee(&self, epoch: EpochId) -> Option<Committee>;
+}
+
+pub trait TransactionStore {
+    fn insert_transaction(&self, transaction: Transaction);
+
+    fn get_transaction(&self, digest: &TransactionDigest) -> Option<Transaction>;
+}
+
+struct PeerHeights {
+    heights: HashMap<PeerId, Option<CheckpointSequenceNumber>>,
+    unprocessed_checkpoints: HashMap<CheckpointDigest, Checkpoint>,
+    sequence_number_to_digest: HashMap<CheckpointSequenceNumber, HashSet<CheckpointDigest>>,
+}
+
+impl PeerHeights {
+    pub fn highest_known_checkpoint(&self) -> Option<&Checkpoint> {
+        self.heights
+            .values()
+            .max()
+            .and_then(Clone::clone)
+            .and_then(|s| {
+                self.sequence_number_to_digest.get(&s).map(|digest_set| {
+                    digest_set
+                        .iter()
+                        .next()
+                        .expect("this should never be empty")
+                })
+            })
+            .and_then(|digest| self.unprocessed_checkpoints.get(digest))
+    }
+
+    pub fn update_peer_height(&mut self, peer_id: PeerId, checkpoint: Option<Checkpoint>) {
+        use std::collections::hash_map::Entry;
+
+        let latest = checkpoint
+            .as_ref()
+            .map(|checkpoint| checkpoint.sequence_number());
+
+        match self.heights.entry(peer_id) {
+            Entry::Occupied(mut entry) => {
+                if latest > *entry.get() {
+                    entry.insert(latest);
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(latest);
+            }
+        }
+
+        if let Some(checkpoint) = checkpoint {
+            self.insert_checkpoint(checkpoint);
+        }
+    }
+
+    pub fn cleanup_old_checkpoints(&mut self, sequence_number: CheckpointSequenceNumber) {
+        self.unprocessed_checkpoints
+            .retain(|_digest, checkpoint| checkpoint.sequence_number() > sequence_number);
+        self.sequence_number_to_digest
+            .retain(|&s, _digests| s > sequence_number);
+    }
+
+    pub fn insert_checkpoint(&mut self, checkpoint: Checkpoint) {
+        let digest = checkpoint.digest();
+        let sequence_number = checkpoint.sequence_number();
+        self.unprocessed_checkpoints.insert(digest, checkpoint);
+        self.sequence_number_to_digest
+            .entry(sequence_number)
+            .or_default()
+            .insert(digest);
+    }
+}
+
+#[derive(Clone, Debug)]
+enum StateSyncMessage {
+    StartSyncJob,
+    // Validators will send this to the StateSyncEventLoop
+    // in order to kick off gossipping it to other peers
+    VerifiedCheckpoint(Box<VerifiedCheckpoint>),
+}
+
+struct StateSyncEventLoop<S> {
+    mailbox: mpsc::Receiver<StateSyncMessage>,
```

Comment:
Can there be a more descriptive name than mailbox? like `rx_commands` or something?

Reply:
imo mailbox is more descriptive and is the term that has generally been used for actors (eg 'mailbox' is the term erlang uses)

Reply:
Just want to push back one more time—whatever the reasoning, I personally find it confusing, and maybe I wouldn't be the only one? We don't use the term mailbox anywhere else in all of sui/narwhal, whether or not it's generally used in other languages? And it doesn't seem all that descriptive to me? Like, a mailbox for what mail? But I'm not gonna bikeshed it more than this, if you feel really strongly.